extract string constants to StartsConstants.java

jkhodadadian opened the pull request:
I extracted almost all of the strings in the project and moved them to the `edu.illinois.starts.constants.StartsConstants` as it was intended. I tried to keep the name of the constants as similar as possible to their value so that the reader can recognize its value from its name.

owolabileg commented on starts-core/src/main/java/edu/illinois/starts/helpers/Cache.java:
```diff
@@ -116,7 +118,7 @@ public void loadM2EdgesFromCache(List<String> moreEdges, String pathString) {
 
     private File createCacheFile(String jar) {
-        String cachePath = jar.replace(m2Repo + File.separator, "").replace(".jar", ".graph");
+        String cachePath = jar.replace(m2Repo + File.separator, BLANK).replace(JAR_TYPE_NAME, GRAPH_TYPE_NAME);
```
`JAR_TYPE_NAME` -> `JAR_EXTENSION`
`GRAPH_TYPE_NAME` -> `GRAPH_EXTENSION`

owolabileg commented on starts-core/src/main/java/edu/illinois/starts/helpers/EkstaziHelper.java:
```diff
@@ -98,18 +100,18 @@ private static boolean isFirstRun(String artifactsDir) {
         Set<String> diffFQNs = new HashSet<>();
         for (String d : diff) {
-            diffFQNs.add(d.replace(".java", "").replace("/", "."));
+            diffFQNs.add(d.replace(JAVA_TYPE_NAME, BLANK).replace(SLASH, DOT));
```
`JAVA_TYPE_NAME` -> `JAVA_EXTENSION`
`SLASH` -> `File.separator`

owolabileg commented on starts-core/src/main/java/edu/illinois/starts/helpers/RTSUtil.java:
```diff
@@ -26,15 +27,15 @@
         Writer.writeGraph(graph, artifactsDir, printGraph, graphFile);
         long end = System.currentTimeMillis();
-        LOGGER.log(Level.FINEST, "[TIME]WRITING FILES: " + (end - start) + "ms");
+        LOGGER.log(Level.FINEST, TIME_WRITING_FILES + (end - start) + MS);
```
`MS` -> `MILLISECOND`

owolabileg commented on starts-core/src/main/java/edu/illinois/starts/helpers/Writer.java:
```diff
@@ -44,7 +46,7 @@ public static void writeToFile(Collection col, String filename, String artifacts
         try (BufferedWriter writer = getWriter(filename)) {
             if (col.isEmpty()) {
-                writer.write("");
+                writer.write(BLANK);
```
`BLANK` -> `EMPTY`

owolabileg commented on starts-core/src/main/java/edu/illinois/starts/helpers/Writer.java:
```diff
@@ -250,6 +252,6 @@ public static String getJarToChecksumMapping(String jar) {
     public static String urlToFQN(String url) {
         // ASSUMPTION: "classes/" rarely occurs in the rest of the path
-        return url.split("classes" + File.separator)[1].replace(".class", "").replace(File.separator, ".");
+        return url.split(CLASSES + File.separator)[1].replace(CLASS_TYPE_NAME, BLANK).replace(File.separator, DOT);
```
`CLASS_TYPE_NAME` -> `CLASS_EXTENSION`

owolabileg commented on starts-core/src/main/java/edu/illinois/starts/helpers/YasglHelper.java:
```diff
@@ -81,14 +83,14 @@ public static void internAndAddEdge(DirectedGraphBuilder<String> builder, String
     @SuppressWarnings("checkstyle:Regexp")
     public DirectedGraphBuilder<String> addEdgesToBuilder(File graphFile, DirectedGraphBuilder<String> builder) {
-        boolean noGZ = graphFile.getAbsolutePath().endsWith(".gz") ? false : true;
+        boolean noGZ = graphFile.getAbsolutePath().endsWith(GZ_TYPE_NAME) ? false : true;
```
`GZ_TYPE_NAME` -> `GZ_EXTENSION`

owolabileg commented on starts-core/src/main/java/edu/illinois/starts/maven/AbstractMojoInterceptor.java:
```diff
@@ -29,7 +29,7 @@ public static URL extractJarURL(Class<?> clz) throws IOException {
 
     public static URL getResource(Class<?> clz) {
-        URL resource = clz.getResource("/" + clz.getName().replace('.', File.separatorChar) + ".class");
+        URL resource = clz.getResource(SLASH + clz.getName().replace('.', File.separatorChar) + CLASS_TYPE_NAME);
```
Does `File.separator()` work here instead of `SLASH`, on both UNIX and Windows?

owolabileg commented on starts-core/src/main/java/edu/illinois/starts/util/ChecksumUtil.java:
```diff
@@ -99,7 +101,7 @@ public static void saveCheckSums(Map<String, Set<RegData>> newCheckSums, String
 
     public static String makeCheckSumPath(String test, String artifactsDir) {
-        return artifactsDir + test + ".clz";
+        return artifactsDir + test + CLZ_TYPE_NAME;
```
`CLZ_TYPE_NAME` -> `CLZ_EXTENSION`

owolabileg commented:
@jkhodadadian, thanks! This is quite thorough in addressing #24. I have left some comments. Please also make the following changes:

1. do not refactor classes in `starts-core/src/main/java/edu/illinois/starts/asm`
2. If a string appears only once in the codebase, do not refactor 
3. If all the multiple (2 or more) occurrences of a string are in the same file, replace with `static final` string in the same class, rather than in `STARTS_CONSTANTS`. In other words, put strings in `edu.illinois.starts.constants.StartsConstants` that are used in multiple files

owolabileg commented:
@jkhodadadian I forgot to mention: either squash your commits into one or use distinct and expressive commit messages.

jkhodadadian commented:
@owolabileg excuse me for the wrong second commit message! it is just some minor changes to fix some checkstyle errors!

owolabileg commented on starts-core/src/main/java/edu/illinois/starts/asm/ClassReader.java:
```diff
@@ -42,7 +43,7 @@
  */
-public class ClassReader {
+public class ClassReader implements StartsConstants {
 
```
Why implement `StartsConstants` if you will not use anything from that interface? Also do not extract constants from any ASM class.

jkhodadadian replied:
I didn't remove `implements StartsConstants` and `import StartsConstants` from ASM classes because I thought you might want to extract the strings in these classes in future. In next commit, I will remove them all. Also, none of ASM classes' constants are extracted.

owolabileg commented on starts-core/src/main/java/edu/illinois/starts/asm/ClassWriter.java:
```diff
@@ -39,7 +40,7 @@
  * @author Eric Bruneton
  */
-public class ClassWriter extends ClassVisitor {
+public class ClassWriter extends ClassVisitor implements StartsConstants {
```
Same for above ASM class

owolabileg commented on starts-core/src/main/java/edu/illinois/starts/asm/FieldWriter.java:
```diff
@@ -28,14 +28,15 @@
  * @author Eric Bruneton
  */
-final class FieldWriter extends FieldVisitor {
+final class FieldWriter extends FieldVisitor implements StartsConstants {
```
Same for above ASM class

owolabileg commented on starts-core/src/main/java/edu/illinois/starts/asm/Frame.java:
```diff
@@ -28,14 +28,15 @@
  * @author Eric Bruneton
  */
-final class Frame {
+final class Frame implements StartsConstants {
```
same for above ASM class

owolabileg commented on starts-core/src/main/java/edu/illinois/starts/asm/Handle.java:
```diff
@@ -37,7 +38,7 @@
  * @author Eric Bruneton
  */
-public final class Handle {
+public final class Handle implements StartsConstants {
```
same for above ASM class

owolabileg commented on starts-core/src/main/java/edu/illinois/starts/asm/Label.java:
```diff
@@ -39,7 +40,7 @@
  * @author Eric Bruneton
  */
-public class Label {
+public class Label implements StartsConstants {
```
same for above ASM class

owolabileg commented on starts-core/src/main/java/edu/illinois/starts/asm/MethodWriter.java:
```diff
@@ -38,7 +39,7 @@
  * @author Eugene Kuleshov
  */
-class MethodWriter extends MethodVisitor {
+class MethodWriter extends MethodVisitor implements StartsConstants {
```
same for above ASM class

owolabileg commented on starts-core/src/main/java/edu/illinois/starts/asm/Type.java:
```diff
@@ -40,7 +42,7 @@
  * @author Chris Nokleberg
  */
-public class Type {
+public class Type implements StartsConstants {
```
same for above ASM class

owolabileg commented on starts-core/src/main/java/edu/illinois/starts/helpers/PomUtil.java:
```diff
@@ -51,7 +51,7 @@ public static String extractParamValue(Plugin plugin, String elem) throws MojoEx
 
     public static Plugin getSfPlugin(MavenProject project) throws MojoExecutionException {
-        Plugin sfPlugin = lookupPlugin("org.apache.maven.plugins:maven-surefire-plugin", project);
+        Plugin sfPlugin = lookupPlugin(SUREFIRE_PLUGIN, project);
```
`org.apache.maven.plugins:maven-surefire-plugin` is used only in one place; no need to extract.

owolabileg commented on starts-core/src/main/java/edu/illinois/starts/helpers/PomUtil.java:
```diff
@@ -64,7 +64,7 @@ public static void checkSFVersion(Plugin sfPlugin) throws MojoExecutionException
         String version = sfPlugin.getVersion();
         if (StartsConstants.MIN_SUREFIRE_VERSION.compareTo(version) > 0) {
```
`StartsConstants.MIN_SUREFIRE_VERSION` --> `MIN_SUREFIRE_VERSION`

owolabileg replied:
This `MIN_SUREFIRE_VERSION` should be a class constant because it's only used in this one class

owolabileg commented on starts-core/src/main/java/edu/illinois/starts/helpers/PomUtil.java:
```diff
@@ -64,7 +64,7 @@ public static void checkSFVersion(Plugin sfPlugin) throws MojoExecutionException
         if (StartsConstants.MIN_SUREFIRE_VERSION.compareTo(version) > 0) {
             throw new MojoExecutionException("Unsupported Surefire version: " + version
-                    + ". Use version " + StartsConstants.MIN_SUREFIRE_VERSION + " and above.");
+                    + ". Use version " + MIN_SUREFIRE_VERSION + " and above.");
```
Good catch

owolabileg commented on starts-core/src/main/java/edu/illinois/starts/maven/AgentLoader.java:
```diff
@@ -76,22 +80,22 @@ private static String getPID() {
 
     private static URL findToolsJar() throws MalformedURLException {
-        String javaHome = System.getProperty("java.home");
+        String javaHome = System.getProperty(JAVA_DOT_HOME);
```
JAVA_DOT_HOME --> JAVA_HOME

owolabileg commented on starts-core/src/main/java/edu/illinois/starts/constants/StartsConstants.java:
```diff
@@ -9,26 +9,261 @@
+
 public interface StartsConstants {
-    String STARTS_DIRECTORY_PATH = ".starts" + File.separator;
+
```
many of the string constants in this file are not used elsewhere. Please remove all unused constants.

jkhodadadian replied:
Here too! In case they will be used again in different files, I didn't remove them. I will do it as you requested. Also because you had some constants in initial `StartsConstants.java` (at the time I forked the project) that were used only one time in one file (like `SUREFIRE_INTERCEPTOR_CLASS_VM`), I thought it's okay. But now I will apply your criteria to all strings.

owolabileg commented on starts-core/src/main/java/edu/illinois/starts/asm/Type.java:
```diff
@@ -33,6 +33,7 @@
 import java.lang.reflect.Constructor;
 import java.lang.reflect.Method;
 
+
```
remove this empty line

owolabileg commented on starts-core/src/main/java/edu/illinois/starts/data/ZLCData.java:
```diff
@@ -10,10 +10,13 @@
 
+import edu.illinois.starts.constants.StartsConstants;
+
+
```
two unecessary empty lines

jkhodadadian replied:
No. only one of them is redundant. One is needed for checkstyle.

owolabileg commented on starts-core/src/main/java/edu/illinois/starts/helpers/EkstaziHelper.java:
```diff
@@ -15,14 +15,16 @@
 import edu.illinois.starts.util.Pair;
 import org.ekstazi.check.AffectedChecker;
 
+
```
unecessary empty line

owolabileg commented on starts-core/src/main/java/edu/illinois/starts/helpers/Writer.java:
```diff
@@ -25,26 +25,29 @@
 import edu.illinois.yasgl.Edge;
 import org.apache.commons.codec.binary.Hex;
 
+
```
unecessary empty line

owolabileg commented on starts-core/src/main/java/edu/illinois/starts/maven/AgentLoader.java:
```diff
@@ -7,20 +7,24 @@
 
+import edu.illinois.starts.constants.StartsConstants;
+
+
```
unnecessary empty line

owolabileg commented on starts-core/src/main/java/edu/illinois/starts/util/Logger.java:
```diff
@@ -10,7 +10,10 @@
-public class Logger {
+import edu.illinois.starts.constants.StartsConstants;
+
+
```
unnecessary empty line

owolabileg commented:
1. you added a bunch of empty lines; please remove them
2. you inlined a number of constants that were already in `StartsConstants.java` before you started this PR; please revert those to be as they were
3.

jkhodadadian commented:
@owolabileg 
1. I removed all of them, whether those you mentioned or those in other classes you didn't mention.
2. I reverted those constants as they were before this PR except the `MIN_SUREFIRE_VERSION` that you said to put it in `PomUtil` class.

owolabileg commented on starts-plugin/src/main/java/edu/illinois/starts/jdeps/RunMojo.java:
```diff
@@ -164,8 +166,8 @@ private boolean hasSameJarChecksum(String cleanSfClassPath) throws MojoExecution
         StringBuilder sb = new StringBuilder();
-        String classes = File.separator + "target" +  File.separator + "classes";
-        String testClasses = File.separator + "target" + File.separator + "test-classes";
+        String classes = FILE_SEPARATOR + TARGET +  FILE_SEPARATOR + CLASSES;
```
Please keep `File.separator` inlined, rather than as a variable.
Once this is fixed, this PR should be ready to merge.
